fdbserver/LogSystem.h

```diff
@@ -766,12 +768,16 @@ struct LogPushData : NonCopyable {
 	void addMessage( StringRef rawMessageWithoutLength, bool usePreviousLocations = false ) {
 		if( !usePreviousLocations ) {
 			prev_tags.clear();
-			if(logSystem->hasRemoteLogs()) {
-				prev_tags.push_back( logSystem->getRandomRouterTag() );
-			}
 			for(auto& tag : next_message_tags) {
 				prev_tags.push_back(tag);
 			}
+			if(logSystem->hasRemoteLogs()) {
+				int primaryLogs = logSystem->getLogSystemConfig().tLogs[0].tLogs.size();
+				int logRouterTags = logSystem->getLogRouterTags();
+				int nWays = logRouterTags / primaryLogs;
+				int newId = (deterministicRandom()->randomChoice(prev_tags).id % primaryLogs) * deterministicRandom()->randomInt(1, nWays+1);
+				prev_tags.push_back( Tag(tagLocalityLogRouter, newId) );
+			}
 			msg_locations.clear();
 			logSystem->getPushLocations( prev_tags, msg_locations );
 			next_message_tags.clear();
```

Comment:
It was like this always, but most of the lines in this block are common with `addTypedMessage`, so perhaps now is a good time to make it private method called `updateMessageLocations` -- or something.